src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -297,168 +354,117 @@ public override Stream EndGetRequestStream(IAsyncResult asyncResult)
-            }
-#endif
+            return headers.ToDictionary((h1) => h1.Key, (h2) => string.Join(",", h2.Value));
         }
 
-        /// <summary>
-        /// Attempts to set the UserAgent property if it exists other wise returns false
-        /// </summary>
-        /// <param name="httpWebRequest">The http web request.</param>
-        /// <param name="headerValue">The value of the user agent.</param>
-        internal static void SetUserAgentHeader(HttpWebRequest httpWebRequest, string headerValue)
+        /// <summary> Returns the underlying HttpRequestMessage </summary>
+        public HttpRequestMessage HttpRequestMessage
         {
-#if !PORTABLELIB
-            httpWebRequest.UserAgent = headerValue;
-#endif
-#if PORTABLELIB
-            if (IsRunningOnSilverlight || !SettingUserAgentPropertyValidOnCurrentPlatform)
+            get
             {
-                return;
+                return _requestMessage;
             }
+        }
 
-            bool headerSet = false;
+        private static HttpWebResponseMessage ConvertHttpWebResponse(HttpResponseMessage response)
+        {
+            IEnumerable<KeyValuePair<string, string>> allHeaders = HttpHeadersToStringDictionary(response.Headers).Concat(HttpHeadersToStringDictionary(response.Content.Headers));
+            return new HttpWebResponseMessage(
+                allHeaders.ToDictionary((h1) => h1.Key, (h2) => h2.Value),
+                (int)response.StatusCode,
+                () =>
+                {
+                    Task<Stream> task = response.Content.ReadAsStreamAsync();
+                    task.Wait();
+                    return task.Result;
+                });
+        }
 
-            // Weird behavior, looks like this won't work in silverlight on the .net portable libaries.
-            // http://social.msdn.microsoft.com/Forums/en-US/netfxbcl/thread/7440fc9d-3490-4823-ac93-73706b535c6a/
-            // http://connect.microsoft.com/VisualStudio/feedback/details/770104/cannot-set-useragent-of-a-httpwebrequest-in-portable-class-and-winrt-libraries
-            // On Windows 8 store throws InvalidOperationException when setting the UserAgent
-            if (UserAgentMethodSetter != null)
+        private static T UnwrapAggregateException<T>(Func<T> action)
+        {
+            try
             {
-                try
-                {
-                    UserAgentMethodSetter.Invoke(httpWebRequest, new object[] { headerValue });
-                    headerSet = true;
-                }
-                catch (InvalidOperationException)
-                {
-                    SettingUserAgentPropertyValidOnCurrentPlatform = false;
-                }
-                catch (TargetInvocationException)
+                return action();
+            }
+            catch (AggregateException aggregateException)
+            {
+                if (aggregateException.InnerExceptions.Count == 1)

```

Comment:
@ElizabethOkerio Wouldn't it be safer to use `>= 1` or are we always guaranteed a single exception in `InnerExceptions`?

Reply:
argh, never mind. Just noticed there's a throw in the enclosing block that will take of things if we have more than 1 exceptions